### PR TITLE
Add new abstract file I/O APIs

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
Fixes https://github.com/ballerina-platform/module-ballerina-io/issues/21

## Approach
Implemented another abstract File IO API layer which helps to directly execute read and write operations on files(bytes, string, JSON, XML, and CSV).

## Release note
Enable read and write abstract file I/O APIs on these following files
- bytes
- string
- JSON
- XML
- CSV

## Automation tests
 - Unit tests 
   > Yes

## Samples
- https://github.com/ballerina-platform/module-ballerina-io/tree/ccde91a67ea5aac5ecd9e03a8e5fd6ac1e2940e9/io-ballerina/src/nio
- Check API docs

## Test environment
- macOS
- JDK 11
- SLP5

### CODE REVIEW

File: io-native/src/main/java/org/ballerinalang/stdlib/io/file/FileIO.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+
```
Comment (kalaiyarasiganeshalingam) [inline io-native/src/main/java/org/ballerinalang/stdlib/io/file/FileIO.java:34]: Missing a Javadoc comment. Check other places as well.

Comment (BuddhiWathsala) [reply io-native/src/main/java/org/ballerinalang/stdlib/io/file/FileIO.java:34]: Fixed in here https://github.com/ballerina-platform/module-ballerina-io/pull/26/commits/3fbc359e6da0a46235bc5295c57e58706235e10d

File: io-ballerina/src/io/character/file_xml_io.bal [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# + return - Either a XML or `io:Error`
+public function fileReadXml(@untainted string path) returns @tainted readonly & xml|Error {
+	var fileOpenResult = openReadableFile(path);
+	if (fileOpenResult is ReadableByteChannel) {
+		ReadableCharacterChannel readableCharacterChannel = new (fileOpenResult, DEFAULT_ENCODING);
+        var fileReadResult = readableCharacterChannel.readXml();
+        var closeResult = fileOpenResult.close();
+        if (closeResult is Error) {
+            return closeResult;
+        }
```
Comment (niveathika) [inline io-ballerina/src/io/character/file_xml_io.bal:31]: In this case, however, even though close fails, the file is read isn't it?

Comment (BuddhiWathsala) [reply io-ballerina/src/io/character/file_xml_io.bal:31]: Yes.

Comment (niveathika) [reply io-ballerina/src/io/character/file_xml_io.bal:31]: In that case, don't we need to return that?

Comment (BuddhiWathsala) [reply io-ballerina/src/io/character/file_xml_io.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-io/pull/26/commits/0b20a7ab26517624cff62e15f049cacac770151e

File: io-ballerina/src/io/bytes/file_byte_io.bal [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/java;
+
+# Read the entire file content as a byte array.
+# ```ballerina
+# byte[]|io:Error content = io:fileReadBytes("./resources/myfile.txt");
+# ```
+# + path - File path
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/bytes/file_byte_io.bal:23]: Shall we specify whether it is `relative-path` or `absolute-path` or either?

File: io-ballerina/src/io/record/file_csv_io.bal [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+# + return - Either an array of string arrays or `io:Error`
+public function fileReadCsv(@untainted string path,
+                        Separator fieldSeparator = ",",
+                        int skipHeaders = 0) returns @tainted readonly & string[][]|Error {
+	
+	var csvFileOpenResult = openReadableCsvFile(path, fieldSeparator, DEFAULT_ENCODING, skipHeaders);
+	if (csvFileOpenResult is ReadableCSVChannel) {
+	    string[][] results = [];
+	    int i = 0;
+
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/record/file_csv_io.bal:33]:
```suggestion
```

File: io-ballerina/src/io/tests/unit-tests/bytes_io.bal [hunk 5]
```diff
@@ -73,6 +72,33 @@ function testWriteBytes() {
     closeWritableBytesChannel();
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/bytes_io.bal:72]: shouldn't this `test:assertFail`

File: io-ballerina/src/io/tests/unit-tests/bytes_io.bal [hunk 6]
```diff
@@ -73,6 +72,33 @@ function testWriteBytes() {
...
+    string content = "Sheldon Cooper";
+    var result = fileWriteBytes(filePath, content.toBytes());
+    
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
+@test:Config {
+    dependsOn: ["testFileWriteBytes"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/bytes_io.bal:88]: IMO, `testFileReadBytes` doesn't have to depend on `testFileWriteBytes` unless testFileReadBytes` needs to read bytes from the file generated by testFileWriteBytes` previously. If that's the case isn't it better to read from a resource file and remove the dependency on the other test.

File: io-ballerina/src/io/tests/unit-tests/char_io.bal [hunk 7]
```diff
@@ -201,9 +199,51 @@ function testReadJson() {
...
+    };
+    var result = fileWriteJson(filePath, content);
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
 @test:Config {
-    dependsOn: ["testReadJson"]
+    dependsOn: ["testFileWriteJson"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/char_io.bal:223]: Same here.

File: io-ballerina/src/io/tests/unit-tests/char_io.bal [hunk 8]
```diff
@@ -324,9 +362,81 @@ function testReadXml() {
...
+                   </CATALOG>`;
+    var result = fileWriteXml(filePath, content);
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
 @test:Config {
-    dependsOn: ["testReadXml"]
+    dependsOn: ["testFileWriteXml"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/char_io.bal:401]: Same here.

File: io-ballerina/src/io/tests/unit-tests/char_io.bal [hunk 9]
```diff
@@ -384,6 +494,60 @@ function testWriteProperties() {
...
+    string filePath = TEMP_DIR + "stringContent.txt";
+    string content = "The Big Bang Theory";
+    var result = fileWriteString(filePath, content);
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
+@test:Config {
+    dependsOn: ["testFileWriteString"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/char_io.bal:508]: Same here.

File: io-ballerina/src/io/tests/unit-tests/char_io.bal [hunk 10]
```diff
@@ -384,6 +494,60 @@ function testWriteProperties() {
...
+    string[] content = ["The Big Bang Theory", "F.R.I.E.N.D.S",
+                        "Game of Thrones", "LOST"];
+    var result = fileWriteLines(filePath, content);
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
+@test:Config {
+    dependsOn: ["testFileWriteLines"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/char_io.bal:533]: Same here.

File: io-ballerina/src/io/tests/unit-tests/csv_io.bal [hunk 11]
```diff
@@ -654,6 +654,126 @@ function testTableWithHeader() {
...
+    ];
+    string filePath = TEMP_DIR + "workers1.csv";
+    var result = fileWriteCsv(filePath, content);
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
+@test:Config{
+    dependsOn: ["testFileCsvWrite"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/csv_io.bal:672]: Same here.

File: io-ballerina/src/io/tests/unit-tests/char_io.bal [hunk 12]
```diff
@@ -282,7 +320,7 @@ function testWriteXml() {
     dependsOn: ["testWriteXml"]
 }
 function testReadXml() {
-    string filePath = TEMP_DIR + "xmlCharsFile.xml";
+    string filePath = TEMP_DIR + "xmlCharsFile1.xml";
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/char_io.bal:323]: Reading an empty XML file?

File: io-ballerina/src/io/tests/unit-tests/char_io.bal [hunk 13]
```diff
@@ -282,7 +320,7 @@ function testWriteXml() {
     dependsOn: ["testWriteXml"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/char_io.bal:320]: Same here.

File: io-ballerina/src/io/tests/unit-tests/csv_io.bal [hunk 14]
```diff
@@ -654,6 +654,126 @@ function testTableWithHeader() {
...
+    ];
+    string filePath = TEMP_DIR + "workers3.csv";
+    var result = fileWriteCsv(filePath, content, COLON);
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
+@test:Config{
+    dependsOn: ["testFileCsvWriteWithColon"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/csv_io.bal:752]: Same here.

File: io-ballerina/src/io/tests/unit-tests/csv_io.bal [hunk 15]
```diff
@@ -654,6 +654,126 @@ function testTableWithHeader() {
...
+    ];
+    string filePath = TEMP_DIR + "workers2.csv";
+    var result = fileWriteCsv(filePath, content, COMMA, 1);
+    if (result is Error) {
+        test:assertFail(msg = result.message());
+    }
+}
+
+@test:Config{
+    dependsOn: ["testFileCsvWriteWithSkipHeaders"]
```
Comment (MadhukaHarith92) [inline io-ballerina/src/io/tests/unit-tests/csv_io.bal:712]: Same here.

**Comment by BuddhiWathsala** [comment]
The required suggestions and the implementation added here https://github.com/ballerina-platform/module-ballerina-io/pull/31.

**Pull request closed without merging**